// @struktoai/mirage-node/dist/chunk-HEQL3LPK.js
import {
  redactSupabaseConfig,
  supabaseToS3Config
} from "./chunk-ALAMUIGJ.js";
import {
  S3Resource
} from "./chunk-G7O6DAWU.js";

// src/resource/supabase/supabase.ts
import { ResourceName } from "@struktoai/mirage-core";

// src/resource/supabase/prompt.ts
var SUPABASE_PROMPT = `{prefix}
  Remote Supabase Storage bucket.
  IMPORTANT: This is a remote mount. Prefer targeted reads over full scans.
  Supports: ls, cat, head, tail, grep, wc, find, stat.`;

// src/resource/supabase/supabase.ts
function remapOpsResource(ops, to) {
  return ops.map((op) => op.resource === ResourceName.S3 ? { ...op, resource: to } : op);
}
function remapCommandsResource(commands, to) {
  return commands.map((cmd) => {
    if (cmd.resource !== ResourceName.S3) return cmd;
    const proto = Object.getPrototypeOf(cmd);
    const copy = Object.assign(Object.create(proto), cmd);
    Object.assign(copy, { resource: to });
    return copy;
  });
}
var SupabaseResource = class extends S3Resource {
  kind = ResourceName.SUPABASE;
  prompt = SUPABASE_PROMPT;
  supabaseConfig;
  supabaseOps;
  supabaseCommands;
  constructor(config) {
    super(supabaseToS3Config(config));
    this.supabaseConfig = config;
    this.supabaseOps = remapOpsResource(super.ops(), ResourceName.SUPABASE);
    this.supabaseCommands = remapCommandsResource(super.commands(), ResourceName.SUPABASE);
  }
  ops() {
    return this.supabaseOps;
  }
  commands() {
    return this.supabaseCommands;
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["accessKeyId", "secretAccessKey", "sessionToken"],
      config: redactSupabaseConfig(this.supabaseConfig)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  SUPABASE_PROMPT,
  SupabaseResource
};
