// @struktoai/mirage-node/dist/chunk-WSDQ2MRG.js
import {
  redactDiscordConfig
} from "./chunk-XVYZO2FG.js";

// src/resource/discord/discord.ts
import {
  DISCORD_API,
  DISCORD_COMMANDS,
  DISCORD_PROMPT,
  DISCORD_VFS_OPS,
  DISCORD_WRITE_PROMPT,
  DiscordAccessor,
  HttpDiscordTransport,
  PathSpec,
  RAMIndexCacheStore,
  ResourceName,
  discordRead,
  discordReaddir,
  discordStat,
  resolveDiscordGlob
} from "@struktoai/mirage-core";
var NodeDiscordTransport = class extends HttpDiscordTransport {
  constructor(token) {
    super();
    this.token = token;
  }
  token;
  baseUrl() {
    return DISCORD_API;
  }
  authHeaders() {
    return { Authorization: `Bot ${this.token}` };
  }
};
var DiscordResource = class {
  kind = ResourceName.DISCORD;
  isRemote = true;
  indexTtl = 600;
  prompt = DISCORD_PROMPT;
  writePrompt = DISCORD_WRITE_PROMPT;
  config;
  accessor;
  index;
  constructor(config) {
    this.config = config;
    this.accessor = new DiscordAccessor(new NodeDiscordTransport(config.token));
    this.index = new RAMIndexCacheStore({ ttl: this.indexTtl });
  }
  open() {
    return Promise.resolve();
  }
  close() {
    return Promise.resolve();
  }
  commands() {
    return DISCORD_COMMANDS;
  }
  ops() {
    return DISCORD_VFS_OPS;
  }
  readFile(p) {
    return discordRead(this.accessor, p, this.index);
  }
  readdir(p) {
    return discordReaddir(this.accessor, p, this.index);
  }
  stat(p) {
    return discordStat(this.accessor, p, this.index);
  }
  async fingerprint(p) {
    const lookup = await this.index.get(p.original);
    return lookup.entry?.remoteTime ?? null;
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return resolveDiscordGlob(this.accessor, effective, this.index);
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["token"],
      config: redactDiscordConfig(this.config)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  DiscordResource
};
