// @struktoai/mirage-node/dist/chunk-H3JR7MN3.js
// src/resource/gcs/config.ts
import { normalizeFields } from "@struktoai/mirage-core";
var GCS_ENDPOINT = "https://storage.googleapis.com";
function gcsToS3Config(config) {
  return {
    bucket: config.bucket,
    region: config.region ?? "auto",
    endpoint: config.endpoint ?? GCS_ENDPOINT,
    accessKeyId: config.accessKeyId,
    secretAccessKey: config.secretAccessKey,
    ...config.timeoutMs !== void 0 ? { timeoutMs: config.timeoutMs } : {}
  };
}
function redactGcsConfig(config) {
  return {
    bucket: config.bucket,
    accessKeyId: "<REDACTED>",
    secretAccessKey: "<REDACTED>",
    endpoint: config.endpoint ?? GCS_ENDPOINT,
    region: config.region ?? "auto",
    ...config.timeoutMs !== void 0 ? { timeoutMs: config.timeoutMs } : {}
  };
}
function normalizeGcsConfig(input) {
  return normalizeFields(input, {
    rename: {
      access_key_id: "accessKeyId",
      secret_access_key: "secretAccessKey",
      endpoint_url: "endpoint",
      timeout: "timeoutMs"
    },
    transform: {
      timeout: (v) => typeof v === "number" ? v * 1e3 : v
    },
    drop: ["proxy"]
  });
}

export {
  GCS_ENDPOINT,
  gcsToS3Config,
  redactGcsConfig,
  normalizeGcsConfig
};
