// @struktoai/mirage-node/dist/chunk-3PBSL7BS.js
import {
  redactGSheetsConfig
} from "./chunk-I4RRGZGB.js";

// src/resource/gsheets/gsheets.ts
import {
  GSHEETS_COMMANDS,
  GSHEETS_PROMPT,
  GSHEETS_VFS_OPS,
  GSHEETS_WRITE_PROMPT,
  GSheetsAccessor,
  PathSpec,
  RAMIndexCacheStore,
  ResourceName,
  TokenManager,
  gsheetsRead,
  gsheetsReaddir,
  gsheetsResolveGlob,
  gsheetsStat
} from "@struktoai/mirage-core";
var GSheetsResource = class {
  kind = ResourceName.GSHEETS;
  isRemote = true;
  indexTtl = 86400;
  prompt = GSHEETS_PROMPT;
  writePrompt = GSHEETS_WRITE_PROMPT;
  config;
  accessor;
  index;
  constructor(config) {
    this.config = config;
    const tm = new TokenManager({
      clientId: config.clientId,
      clientSecret: config.clientSecret,
      refreshToken: config.refreshToken
    });
    this.accessor = new GSheetsAccessor({ tokenManager: tm });
    this.index = new RAMIndexCacheStore({ ttl: 86400 });
  }
  open() {
    return Promise.resolve();
  }
  close() {
    return Promise.resolve();
  }
  commands() {
    return GSHEETS_COMMANDS;
  }
  ops() {
    return GSHEETS_VFS_OPS;
  }
  readFile(p) {
    return gsheetsRead(this.accessor, p, this.index);
  }
  readdir(p) {
    return gsheetsReaddir(this.accessor, p, this.index);
  }
  stat(p) {
    return gsheetsStat(this.accessor, p, this.index);
  }
  async fingerprint(p) {
    const lookup = await this.index.get(p.original);
    return lookup.entry?.remoteTime ?? null;
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return gsheetsResolveGlob(this.accessor, effective, this.index);
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["clientSecret", "refreshToken"],
      config: redactGSheetsConfig(this.config)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  GSheetsResource
};
