// @struktoai/mirage-node/dist/chunk-6M2NL363.js
import {
  redactBoxConfig
} from "./chunk-EKPH4MMD.js";

// src/resource/box/box.ts
import {
  BOX_COMMANDS,
  BOX_PROMPT,
  BOX_VFS_OPS,
  BoxAccessor,
  BoxTokenManager,
  PathSpec,
  RAMIndexCacheStore,
  ResourceName,
  boxRead,
  boxReaddir,
  boxResolveGlob,
  boxStat
} from "@struktoai/mirage-core";
var BoxResource = class {
  kind = ResourceName.BOX;
  isRemote = true;
  indexTtl = 86400;
  prompt = BOX_PROMPT;
  config;
  accessor;
  index;
  constructor(config) {
    this.config = config;
    const tm = new BoxTokenManager({
      ...config.clientId !== void 0 ? { clientId: config.clientId } : {},
      ...config.clientSecret !== void 0 ? { clientSecret: config.clientSecret } : {},
      ...config.refreshToken !== void 0 ? { refreshToken: config.refreshToken } : {},
      ...config.accessToken !== void 0 ? { accessToken: config.accessToken } : {},
      ...config.refreshFn !== void 0 ? { refreshFn: config.refreshFn } : {},
      ...config.onRefreshTokenRotated !== void 0 ? { onRefreshTokenRotated: config.onRefreshTokenRotated } : {}
    });
    this.accessor = new BoxAccessor({ tokenManager: tm });
    this.index = new RAMIndexCacheStore({ ttl: 86400 });
  }
  open() {
    return Promise.resolve();
  }
  close() {
    return Promise.resolve();
  }
  commands() {
    return BOX_COMMANDS;
  }
  ops() {
    return BOX_VFS_OPS;
  }
  readFile(p) {
    return boxRead(this.accessor, p, this.index);
  }
  readdir(p) {
    return boxReaddir(this.accessor, p, this.index);
  }
  stat(p) {
    return boxStat(this.accessor, p, this.index);
  }
  async fingerprint(p) {
    const lookup = await this.index.get(p.original);
    return lookup.entry?.remoteTime ?? null;
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return boxResolveGlob(this.accessor, effective, this.index);
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["clientSecret", "refreshToken", "accessToken"],
      config: redactBoxConfig(this.config)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  BoxResource
};
