// @struktoai/mirage-node/dist/chunk-6VKNSO6U.js
// src/resource/r2/config.ts
import { normalizeFields } from "@struktoai/mirage-core";
function resolvedR2Endpoint(config) {
  if (config.endpoint !== void 0 && config.endpoint !== "") return config.endpoint;
  if (config.accountId !== void 0 && config.accountId !== "") {
    return `https://${config.accountId}.r2.cloudflarestorage.com`;
  }
  throw new Error("R2Config requires accountId or endpoint");
}
function r2ToS3Config(config) {
  return {
    bucket: config.bucket,
    region: config.region ?? "auto",
    endpoint: resolvedR2Endpoint(config),
    accessKeyId: config.accessKeyId,
    secretAccessKey: config.secretAccessKey,
    ...config.profile !== void 0 ? { profile: config.profile } : {},
    ...config.timeoutMs !== void 0 ? { timeoutMs: config.timeoutMs } : {}
  };
}
function redactR2Config(config) {
  return {
    bucket: config.bucket,
    accessKeyId: "<REDACTED>",
    secretAccessKey: "<REDACTED>",
    ...config.accountId !== void 0 ? { accountId: config.accountId } : {},
    endpoint: resolvedR2Endpoint(config),
    region: config.region ?? "auto",
    ...config.profile !== void 0 ? { profile: config.profile } : {},
    ...config.timeoutMs !== void 0 ? { timeoutMs: config.timeoutMs } : {}
  };
}
function normalizeR2Config(input) {
  return normalizeFields(input, {
    rename: {
      account_id: "accountId",
      access_key_id: "accessKeyId",
      secret_access_key: "secretAccessKey",
      aws_profile: "profile",
      endpoint_url: "endpoint",
      timeout: "timeoutMs"
    },
    transform: {
      timeout: (v) => typeof v === "number" ? v * 1e3 : v
    },
    drop: ["proxy"]
  });
}

export {
  resolvedR2Endpoint,
  r2ToS3Config,
  redactR2Config,
  normalizeR2Config
};
