// @struktoai/mirage-node/dist/chunk-7QXJ4KWQ.js
// src/resource/mongodb/mongodb.ts
import {
  detectMongoScope,
  MONGODB_COMMANDS,
  MONGODB_OPS,
  MONGODB_PROMPT,
  MongoDBAccessor,
  mongoRead,
  mongoReaddir,
  mongoStat,
  PathSpec,
  RAMIndexCacheStore,
  resolveMongoGlob,
  ResourceName,
  resolveMongoDBConfig
} from "@struktoai/mirage-core";

// src/resource/mongodb/store.ts
import { loadOptionalPeer } from "@struktoai/mirage-core";
var MongoDBStore = class {
  uri;
  clientPromise = null;
  constructor(uri) {
    this.uri = uri;
  }
  async listDatabases() {
    const c = await this._client();
    const r = await c.db().admin().listDatabases();
    return r.databases.map((d) => d.name);
  }
  async listCollections(database) {
    const c = await this._client();
    const cols = await c.db(database).listCollections().toArray();
    return cols.map((col) => col.name);
  }
  async findDocuments(database, collection, filter = {}, options = {}) {
    const c = await this._client();
    let cursor = c.db(database).collection(collection).find(filter, {
      ...options.projection !== void 0 ? { projection: options.projection } : {}
    });
    if (options.sort !== void 0) cursor = cursor.sort(options.sort);
    if (options.skip !== void 0) cursor = cursor.skip(options.skip);
    if (options.limit !== void 0) cursor = cursor.limit(options.limit);
    return await cursor.toArray();
  }
  async countDocuments(database, collection, filter = {}) {
    const c = await this._client();
    return c.db(database).collection(collection).countDocuments(filter);
  }
  async listIndexes(database, collection) {
    const c = await this._client();
    return c.db(database).collection(collection).listIndexes().toArray();
  }
  async close() {
    if (this.clientPromise === null) return;
    const c = await this.clientPromise;
    this.clientPromise = null;
    await c.close();
  }
  async _client() {
    this.clientPromise ??= this._connect();
    return this.clientPromise;
  }
  async _connect() {
    const mod = await loadOptionalPeer(() => import("mongodb"), {
      feature: "MongoDBResource",
      packageName: "mongodb"
    });
    const c = new mod.MongoClient(this.uri);
    await c.connect();
    return c;
  }
};

// src/resource/mongodb/mongodb.ts
var MongoDBResource = class {
  kind = ResourceName.MONGODB;
  isRemote = true;
  indexTtl = 0;
  prompt;
  config;
  store;
  accessor;
  index;
  constructor(options) {
    const { config, prefix } = "config" in options ? options : { config: options, prefix: void 0 };
    this.config = resolveMongoDBConfig(config);
    this.store = new MongoDBStore(this.config.uri);
    this.accessor = new MongoDBAccessor(this.store, this.config);
    this.index = new RAMIndexCacheStore({ ttl: this.indexTtl });
    this.prompt = MONGODB_PROMPT.replace("{prefix}", prefix ?? "");
  }
  open() {
    return Promise.resolve();
  }
  async close() {
    await this.store.close();
  }
  ops() {
    return MONGODB_OPS;
  }
  commands() {
    return MONGODB_COMMANDS;
  }
  readFile(p) {
    return mongoRead(this.accessor, p, this.index);
  }
  readdir(p) {
    return mongoReaddir(this.accessor, p, this.index);
  }
  stat(p) {
    return mongoStat(this.accessor, p, this.index);
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return resolveMongoGlob(this.accessor, effective, this.index);
  }
};

export {
  MongoDBStore,
  MongoDBResource
};
