// @struktoai/mirage-node/dist/chunk-XDEEOTVK.js
import {
  redactGSlidesConfig
} from "./chunk-5ITBZN55.js";

// src/resource/gslides/gslides.ts
import {
  GSLIDES_COMMANDS,
  GSLIDES_PROMPT,
  GSLIDES_VFS_OPS,
  GSLIDES_WRITE_PROMPT,
  GSlidesAccessor,
  PathSpec,
  RAMIndexCacheStore,
  ResourceName,
  TokenManager,
  gslidesRead,
  gslidesReaddir,
  gslidesResolveGlob,
  gslidesStat
} from "@struktoai/mirage-core";
var GSlidesResource = class {
  kind = ResourceName.GSLIDES;
  isRemote = true;
  indexTtl = 86400;
  prompt = GSLIDES_PROMPT;
  writePrompt = GSLIDES_WRITE_PROMPT;
  config;
  accessor;
  index;
  constructor(config) {
    this.config = config;
    const tm = new TokenManager({
      clientId: config.clientId,
      clientSecret: config.clientSecret,
      refreshToken: config.refreshToken
    });
    this.accessor = new GSlidesAccessor({ tokenManager: tm });
    this.index = new RAMIndexCacheStore({ ttl: 86400 });
  }
  open() {
    return Promise.resolve();
  }
  close() {
    return Promise.resolve();
  }
  commands() {
    return GSLIDES_COMMANDS;
  }
  ops() {
    return GSLIDES_VFS_OPS;
  }
  readFile(p) {
    return gslidesRead(this.accessor, p, this.index);
  }
  readdir(p) {
    return gslidesReaddir(this.accessor, p, this.index);
  }
  stat(p) {
    return gslidesStat(this.accessor, p, this.index);
  }
  async fingerprint(p) {
    const lookup = await this.index.get(p.original);
    return lookup.entry?.remoteTime ?? null;
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return gslidesResolveGlob(this.accessor, effective, this.index);
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["clientSecret", "refreshToken"],
      config: redactGSlidesConfig(this.config)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  GSlidesResource
};
