// @struktoai/mirage-node/dist/chunk-G7O6DAWU.js
import {
  redactConfig
} from "./chunk-LPMPTQYP.js";

// src/resource/s3/s3.ts
import {
  copy as copyCore,
  create as createCore,
  du as duCore,
  duAll as duAllCore,
  exists as existsCore,
  find as findCore,
  mkdir as mkdirCore,
  PathSpec,
  RAMIndexCacheStore,
  rangeRead as rangeReadCore,
  S3_COMMANDS,
  read as readCore,
  readdir as readdirCore,
  rename as renameCore,
  ResourceName,
  resolveS3Glob as globCore,
  rmR as rmRCore,
  rmdir as rmdirCore,
  S3_OPS,
  S3_PROMPT,
  S3Accessor,
  stat as statCore,
  stream as streamCore,
  truncate as truncateCore,
  unlink as unlinkCore,
  write as writeCore
} from "@struktoai/mirage-core";
var S3Resource = class {
  kind = ResourceName.S3;
  isRemote = true;
  indexTtl = 600;
  prompt = S3_PROMPT;
  config;
  accessor;
  index;
  opsMap = {
    read_bytes: readCore,
    write: writeCore,
    readdir: readdirCore,
    stat: statCore,
    unlink: unlinkCore,
    rmdir: rmdirCore,
    copy: copyCore,
    rename: renameCore,
    mkdir: mkdirCore,
    read_stream: streamCore,
    range_read: rangeReadCore,
    rm_recursive: rmRCore,
    du_total: duCore,
    du_all: duAllCore,
    create: createCore,
    truncate: truncateCore,
    exists: existsCore,
    find_flat: findCore
  };
  constructor(config) {
    this.config = config;
    this.accessor = new S3Accessor(this.config);
    this.index = new RAMIndexCacheStore({ ttl: this.indexTtl });
  }
  open() {
    return Promise.resolve();
  }
  close() {
    return Promise.resolve();
  }
  commands() {
    return S3_COMMANDS;
  }
  ops() {
    return S3_OPS;
  }
  streamPath(p) {
    return streamCore(this.accessor, p);
  }
  readFile(p) {
    return readCore(this.accessor, p);
  }
  writeFile(p, data) {
    return writeCore(this.accessor, p, data);
  }
  async appendFile(p, data) {
    let existing;
    try {
      existing = await readCore(this.accessor, p);
    } catch (err) {
      if (err?.code === "ENOENT") {
        existing = new Uint8Array();
      } else {
        throw err;
      }
    }
    const merged = new Uint8Array(existing.byteLength + data.byteLength);
    merged.set(existing, 0);
    merged.set(data, existing.byteLength);
    await writeCore(this.accessor, p, merged);
  }
  readdir(p) {
    return readdirCore(this.accessor, p, this.index);
  }
  stat(p) {
    return statCore(this.accessor, p);
  }
  exists(p) {
    return existsCore(this.accessor, p);
  }
  mkdir(p) {
    return mkdirCore(this.accessor, p);
  }
  rmdir(p) {
    return rmdirCore(this.accessor, p);
  }
  unlink(p) {
    return unlinkCore(this.accessor, p);
  }
  rename(src, dst) {
    return renameCore(this.accessor, src, dst);
  }
  truncate(p, length) {
    return truncateCore(this.accessor, p, length);
  }
  copy(src, dst) {
    return copyCore(this.accessor, src, dst);
  }
  rmR(p) {
    return rmRCore(this.accessor, p);
  }
  du(p) {
    return duCore(this.accessor, p);
  }
  find(p, options = {}) {
    return findCore(this.accessor, p, options);
  }
  glob(paths, prefix = "") {
    const effective = prefix ? paths.map(
      (p) => p.prefix ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return globCore(this.accessor, effective, this.index);
  }
  async fingerprint(p) {
    try {
      const s = await statCore(this.accessor, p);
      const etag = s.extra.etag;
      return typeof etag === "string" && etag !== "" ? etag : null;
    } catch (err) {
      if (err?.code === "ENOENT") return null;
      throw err;
    }
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["accessKeyId", "secretAccessKey", "sessionToken"],
      config: redactConfig(this.config)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  S3Resource
};
