// @struktoai/mirage-node/dist/chunk-ALAMUIGJ.js
// src/resource/supabase/config.ts
import { normalizeFields } from "@struktoai/mirage-core";
function resolvedSupabaseEndpoint(config) {
  if (config.endpoint !== void 0 && config.endpoint !== "") return config.endpoint;
  if (config.projectRef !== void 0 && config.projectRef !== "") {
    return `https://${config.projectRef}.storage.supabase.co/storage/v1/s3`;
  }
  throw new Error("SupabaseConfig requires projectRef or endpoint");
}
function supabaseToS3Config(config) {
  return {
    bucket: config.bucket,
    region: config.region,
    endpoint: resolvedSupabaseEndpoint(config),
    accessKeyId: config.accessKeyId,
    secretAccessKey: config.secretAccessKey,
    forcePathStyle: true,
    ...config.sessionToken !== void 0 ? { sessionToken: config.sessionToken } : {},
    ...config.timeoutMs !== void 0 ? { timeoutMs: config.timeoutMs } : {}
  };
}
function redactSupabaseConfig(config) {
  return {
    bucket: config.bucket,
    region: config.region,
    ...config.projectRef !== void 0 ? { projectRef: config.projectRef } : {},
    endpoint: resolvedSupabaseEndpoint(config),
    accessKeyId: "<REDACTED>",
    secretAccessKey: "<REDACTED>",
    ...config.sessionToken !== void 0 ? { sessionToken: "<REDACTED>" } : {},
    ...config.timeoutMs !== void 0 ? { timeoutMs: config.timeoutMs } : {}
  };
}
function normalizeSupabaseConfig(input) {
  return normalizeFields(input, {
    rename: {
      project_ref: "projectRef",
      access_key_id: "accessKeyId",
      secret_access_key: "secretAccessKey",
      session_token: "sessionToken",
      endpoint_url: "endpoint",
      timeout: "timeoutMs"
    },
    transform: {
      timeout: (v) => typeof v === "number" ? v * 1e3 : v
    },
    drop: ["proxy"]
  });
}

export {
  resolvedSupabaseEndpoint,
  supabaseToS3Config,
  redactSupabaseConfig,
  normalizeSupabaseConfig
};
