// @struktoai/mirage-node/dist/chunk-ICSS5ETT.js
import {
  r2ToS3Config,
  redactR2Config
} from "./chunk-6VKNSO6U.js";
import {
  S3Resource
} from "./chunk-G7O6DAWU.js";

// src/resource/r2/r2.ts
import { ResourceName } from "@struktoai/mirage-core";

// src/resource/r2/prompt.ts
var R2_PROMPT = `{prefix}
  Remote Cloudflare R2 bucket.
  IMPORTANT: This is a remote mount. Prefer targeted reads over full scans.
  Supports: ls, cat, head, tail, grep, rg, wc, find, tree, jq, stat.
  cat on .parquet/.orc/.feather returns a formatted table.`;

// src/resource/r2/r2.ts
function remapOpsResource(ops, to) {
  return ops.map((op) => op.resource === ResourceName.S3 ? { ...op, resource: to } : op);
}
function remapCommandsResource(commands, to) {
  return commands.map((cmd) => {
    if (cmd.resource !== ResourceName.S3) return cmd;
    const proto = Object.getPrototypeOf(cmd);
    const copy = Object.assign(Object.create(proto), cmd);
    Object.assign(copy, { resource: to });
    return copy;
  });
}
var R2Resource = class extends S3Resource {
  kind = ResourceName.R2;
  prompt = R2_PROMPT;
  r2Config;
  r2Ops;
  r2Commands;
  constructor(config) {
    super(r2ToS3Config(config));
    this.r2Config = config;
    this.r2Ops = remapOpsResource(super.ops(), ResourceName.R2);
    this.r2Commands = remapCommandsResource(super.commands(), ResourceName.R2);
  }
  ops() {
    return this.r2Ops;
  }
  commands() {
    return this.r2Commands;
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["accessKeyId", "secretAccessKey"],
      config: redactR2Config(this.r2Config)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  R2_PROMPT,
  R2Resource
};
