// @struktoai/mirage-node/dist/chunk-MY3NUJJD.js
import {
  S3Resource
} from "./chunk-G7O6DAWU.js";
import {
  gcsToS3Config,
  redactGcsConfig
} from "./chunk-H3JR7MN3.js";

// src/resource/gcs/gcs.ts
import { ResourceName } from "@struktoai/mirage-core";

// src/resource/gcs/prompt.ts
var GCS_PROMPT = `{prefix}
  Remote Google Cloud Storage bucket.
  IMPORTANT: This is a remote mount. Prefer targeted reads over full scans.
  Supports: ls, cat, head, tail, grep, rg, wc, find, tree, jq, stat.
  cat on .parquet/.orc/.feather returns a formatted table.`;

// src/resource/gcs/gcs.ts
function remapOpsResource(ops, to) {
  return ops.map((op) => op.resource === ResourceName.S3 ? { ...op, resource: to } : op);
}
function remapCommandsResource(commands, to) {
  return commands.map((cmd) => {
    if (cmd.resource !== ResourceName.S3) return cmd;
    const proto = Object.getPrototypeOf(cmd);
    const copy = Object.assign(Object.create(proto), cmd);
    Object.assign(copy, { resource: to });
    return copy;
  });
}
var GCSResource = class extends S3Resource {
  kind = ResourceName.GCS;
  prompt = GCS_PROMPT;
  gcsConfig;
  gcsOps;
  gcsCommands;
  constructor(config) {
    super(gcsToS3Config(config));
    this.gcsConfig = config;
    this.gcsOps = remapOpsResource(super.ops(), ResourceName.GCS);
    this.gcsCommands = remapCommandsResource(super.commands(), ResourceName.GCS);
  }
  ops() {
    return this.gcsOps;
  }
  commands() {
    return this.gcsCommands;
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["accessKeyId", "secretAccessKey"],
      config: redactGcsConfig(this.gcsConfig)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  GCS_PROMPT,
  GCSResource
};
