// @struktoai/mirage-node/dist/chunk-XQUQSNCA.js
// src/resource/postgres/postgres.ts
import {
  PathSpec,
  POSTGRES_COMMANDS,
  POSTGRES_OPS,
  POSTGRES_PROMPT,
  PostgresAccessor,
  postgresRead,
  postgresReaddir,
  postgresStat,
  RAMIndexCacheStore,
  ResourceName,
  resolvePostgresConfig,
  resolvePostgresGlob
} from "@struktoai/mirage-core";

// src/resource/postgres/store.ts
import { loadOptionalPeer } from "@struktoai/mirage-core";
var PostgresStore = class {
  config;
  poolPromise = null;
  constructor(config) {
    this.config = config;
  }
  async query(sql, params) {
    const pool = await this._pool();
    const result = await pool.query(sql, params);
    return { rows: result.rows, rowCount: result.rowCount ?? 0 };
  }
  async databaseName() {
    const result = await this.query("SELECT current_database() AS db");
    return result.rows[0]?.db ?? "";
  }
  async close() {
    if (this.poolPromise === null) return;
    const pool = await this.poolPromise;
    this.poolPromise = null;
    await pool.end();
  }
  async _pool() {
    this.poolPromise ??= this._connect();
    return this.poolPromise;
  }
  async _connect() {
    const mod = await loadOptionalPeer(
      () => import("pg"),
      { feature: "PostgresResource", packageName: "pg" }
    );
    const Pool = mod.default?.Pool ?? mod.Pool;
    if (Pool === void 0) {
      throw new Error("postgres: pg package missing Pool export");
    }
    return new Pool({
      connectionString: this.config.dsn,
      options: "-c default_transaction_read_only=on"
    });
  }
};

// src/resource/postgres/postgres.ts
var PostgresResource = class {
  kind = ResourceName.POSTGRES;
  isRemote = true;
  indexTtl = 0;
  prompt;
  config;
  store;
  accessor;
  index;
  constructor(options) {
    const { config, prefix } = "config" in options ? options : { config: options, prefix: void 0 };
    this.config = resolvePostgresConfig(config);
    this.store = new PostgresStore(this.config);
    this.accessor = new PostgresAccessor(this.store, this.config);
    this.index = new RAMIndexCacheStore({ ttl: this.indexTtl });
    this.prompt = POSTGRES_PROMPT.replace("{prefix}", prefix ?? "");
  }
  open() {
    return Promise.resolve();
  }
  async close() {
    await this.store.close();
  }
  ops() {
    return POSTGRES_OPS;
  }
  commands() {
    return POSTGRES_COMMANDS;
  }
  readFile(p) {
    return postgresRead(this.accessor, p, this.index);
  }
  readdir(p) {
    return postgresReaddir(this.accessor, p, this.index);
  }
  stat(p) {
    return postgresStat(this.accessor, p, this.index);
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return resolvePostgresGlob(this.accessor, effective, this.index);
  }
};

export {
  PostgresStore,
  PostgresResource
};
