// @struktoai/mirage-node/dist/chunk-NOZZFJEB.js
import {
  redactGDocsConfig
} from "./chunk-3Y625AUP.js";

// src/resource/gdocs/gdocs.ts
import {
  GDOCS_COMMANDS,
  GDOCS_PROMPT,
  GDOCS_VFS_OPS,
  GDOCS_WRITE_PROMPT,
  GDocsAccessor,
  PathSpec,
  RAMIndexCacheStore,
  ResourceName,
  TokenManager,
  gdocsRead,
  gdocsReaddir,
  gdocsResolveGlob,
  gdocsStat
} from "@struktoai/mirage-core";
var GDocsResource = class {
  kind = ResourceName.GDOCS;
  isRemote = true;
  indexTtl = 86400;
  prompt = GDOCS_PROMPT;
  writePrompt = GDOCS_WRITE_PROMPT;
  config;
  accessor;
  index;
  constructor(config) {
    this.config = config;
    const tm = new TokenManager({
      clientId: config.clientId,
      clientSecret: config.clientSecret,
      refreshToken: config.refreshToken
    });
    this.accessor = new GDocsAccessor({ tokenManager: tm });
    this.index = new RAMIndexCacheStore({ ttl: 86400 });
  }
  open() {
    return Promise.resolve();
  }
  close() {
    return Promise.resolve();
  }
  commands() {
    return GDOCS_COMMANDS;
  }
  ops() {
    return GDOCS_VFS_OPS;
  }
  readFile(p) {
    return gdocsRead(this.accessor, p, this.index);
  }
  readdir(p) {
    return gdocsReaddir(this.accessor, p, this.index);
  }
  stat(p) {
    return gdocsStat(this.accessor, p, this.index);
  }
  async fingerprint(p) {
    const lookup = await this.index.get(p.original);
    return lookup.entry?.remoteTime ?? null;
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return gdocsResolveGlob(this.accessor, effective, this.index);
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["clientSecret", "refreshToken"],
      config: redactGDocsConfig(this.config)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  GDocsResource
};
