// @struktoai/mirage-node/dist/chunk-DCJQ2PIX.js
import {
  redactLangfuseConfig
} from "./chunk-6M3QVVOH.js";

// src/resource/langfuse/langfuse.ts
import {
  HttpLangfuseTransport,
  LANGFUSE_COMMANDS,
  LANGFUSE_PROMPT,
  LANGFUSE_VFS_OPS,
  LangfuseAccessor,
  langfuseRead,
  langfuseReaddir,
  langfuseStat,
  PathSpec,
  RAMIndexCacheStore,
  ResourceName,
  resolveLangfuseGlob
} from "@struktoai/mirage-core";
var LangfuseResource = class {
  kind = ResourceName.LANGFUSE;
  isRemote = true;
  indexTtl = 600;
  prompt = LANGFUSE_PROMPT;
  config;
  accessor;
  index;
  constructor(config) {
    this.config = config;
    const transportOpts = {
      publicKey: config.publicKey,
      secretKey: config.secretKey
    };
    if (config.host !== void 0) transportOpts.host = config.host;
    const accessorConfig = {};
    if (config.defaultTraceLimit !== void 0) {
      accessorConfig.defaultTraceLimit = config.defaultTraceLimit;
    }
    if (config.defaultSearchLimit !== void 0) {
      accessorConfig.defaultSearchLimit = config.defaultSearchLimit;
    }
    if (config.defaultFromTimestamp !== void 0) {
      accessorConfig.defaultFromTimestamp = config.defaultFromTimestamp;
    }
    this.accessor = new LangfuseAccessor(new HttpLangfuseTransport(transportOpts), accessorConfig);
    this.index = new RAMIndexCacheStore({ ttl: this.indexTtl });
  }
  open() {
    return Promise.resolve();
  }
  close() {
    return Promise.resolve();
  }
  commands() {
    return LANGFUSE_COMMANDS;
  }
  ops() {
    return LANGFUSE_VFS_OPS;
  }
  readFile(p) {
    return langfuseRead(this.accessor, p, this.index);
  }
  readdir(p) {
    return langfuseReaddir(this.accessor, p, this.index);
  }
  stat(p) {
    return langfuseStat(this.accessor, p, this.index);
  }
  fingerprint(_p) {
    return Promise.resolve(null);
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return resolveLangfuseGlob(this.accessor, effective, this.index);
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["secretKey"],
      config: redactLangfuseConfig(this.config)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  LangfuseResource
};
