// @struktoai/mirage-node/dist/chunk-UZZD4MRJ.js
import {
  redactLinearConfig
} from "./chunk-6Y6GWMFF.js";

// src/resource/linear/linear.ts
import {
  HttpLinearTransport,
  LINEAR_COMMANDS,
  LINEAR_PROMPT,
  LINEAR_VFS_OPS,
  LINEAR_WRITE_PROMPT,
  LinearAccessor,
  linearRead,
  linearReaddir,
  linearStat,
  PathSpec,
  RAMIndexCacheStore,
  ResourceName,
  resolveLinearGlob
} from "@struktoai/mirage-core";
var LinearResource = class {
  kind = ResourceName.LINEAR;
  isRemote = true;
  indexTtl = 600;
  prompt = LINEAR_PROMPT;
  writePrompt = LINEAR_WRITE_PROMPT;
  config;
  accessor;
  index;
  constructor(config) {
    this.config = config;
    const transportOpts = { apiKey: config.apiKey };
    if (config.baseUrl !== void 0) transportOpts.baseUrl = config.baseUrl;
    this.accessor = new LinearAccessor(new HttpLinearTransport(transportOpts));
    this.index = new RAMIndexCacheStore({ ttl: this.indexTtl });
  }
  open() {
    return Promise.resolve();
  }
  close() {
    return Promise.resolve();
  }
  commands() {
    return LINEAR_COMMANDS;
  }
  ops() {
    return LINEAR_VFS_OPS;
  }
  filter() {
    const out = {};
    if (this.config.teamIds !== void 0) out.teamIds = this.config.teamIds;
    return out;
  }
  readFile(p) {
    return linearRead(this.accessor, p, this.index, this.filter());
  }
  readdir(p) {
    return linearReaddir(this.accessor, p, this.index, this.filter());
  }
  stat(p) {
    return linearStat(this.accessor, p, this.index);
  }
  async fingerprint(p) {
    const lookup = await this.index.get(p.original);
    return lookup.entry?.remoteTime ?? null;
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return resolveLinearGlob(this.accessor, effective, this.index, this.filter());
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["apiKey"],
      config: redactLinearConfig(this.config)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  LinearResource
};
