// @struktoai/mirage-node/dist/chunk-V5OLWE7F.js
import {
  redactSlackConfig
} from "./chunk-KINMBZHC.js";

// src/resource/slack/slack.ts
import {
  NodeSlackTransport,
  PathSpec,
  RAMIndexCacheStore,
  ResourceName,
  resolveSlackGlob,
  SLACK_COMMANDS,
  SLACK_PROMPT,
  SLACK_VFS_OPS,
  SLACK_WRITE_PROMPT,
  SlackAccessor,
  slackRead,
  slackReaddir,
  slackStat
} from "@struktoai/mirage-core";
var SlackResource = class {
  kind = ResourceName.SLACK;
  isRemote = true;
  indexTtl = 600;
  prompt = SLACK_PROMPT;
  writePrompt = SLACK_WRITE_PROMPT;
  config;
  accessor;
  index;
  constructor(config) {
    this.config = config;
    this.accessor = new SlackAccessor(new NodeSlackTransport(config.token, config.searchToken));
    this.index = new RAMIndexCacheStore({ ttl: this.indexTtl });
  }
  open() {
    return Promise.resolve();
  }
  close() {
    return Promise.resolve();
  }
  commands() {
    return SLACK_COMMANDS;
  }
  ops() {
    return SLACK_VFS_OPS;
  }
  readFile(p) {
    return slackRead(this.accessor, p, this.index);
  }
  readdir(p) {
    return slackReaddir(this.accessor, p, this.index);
  }
  stat(p) {
    return slackStat(this.accessor, p, this.index);
  }
  async fingerprint(p) {
    const lookup = await this.index.get(p.original);
    return lookup.entry?.remoteTime ?? null;
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return resolveSlackGlob(this.accessor, effective, this.index);
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["token", "searchToken"],
      config: redactSlackConfig(this.config)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  SlackResource
};
