// @struktoai/mirage-node/dist/chunk-ZC4KNEOQ.js
// src/resource/oci/config.ts
import { normalizeFields } from "@struktoai/mirage-core";
function resolvedOciEndpoint(config) {
  if (config.endpoint !== void 0 && config.endpoint !== "") return config.endpoint;
  return `https://${config.namespace}.compat.objectstorage.${config.region}.oci.customer-oci.com`;
}
function ociToS3Config(config) {
  return {
    bucket: config.bucket,
    region: config.region,
    endpoint: resolvedOciEndpoint(config),
    accessKeyId: config.accessKeyId,
    secretAccessKey: config.secretAccessKey,
    forcePathStyle: true,
    ...config.timeoutMs !== void 0 ? { timeoutMs: config.timeoutMs } : {}
  };
}
function redactOciConfig(config) {
  return {
    bucket: config.bucket,
    namespace: config.namespace,
    region: config.region,
    accessKeyId: "<REDACTED>",
    secretAccessKey: "<REDACTED>",
    endpoint: resolvedOciEndpoint(config),
    ...config.timeoutMs !== void 0 ? { timeoutMs: config.timeoutMs } : {}
  };
}
function normalizeOciConfig(input) {
  return normalizeFields(input, {
    rename: {
      access_key_id: "accessKeyId",
      secret_access_key: "secretAccessKey",
      endpoint_url: "endpoint",
      timeout: "timeoutMs"
    },
    transform: {
      timeout: (v) => typeof v === "number" ? v * 1e3 : v
    },
    drop: ["proxy"]
  });
}

export {
  resolvedOciEndpoint,
  ociToS3Config,
  redactOciConfig,
  normalizeOciConfig
};
