// @struktoai/mirage-node/dist/chunk-5E6BULAK.js
import {
  redactDropboxConfig
} from "./chunk-RQYYYBQG.js";

// src/resource/dropbox/dropbox.ts
import {
  DROPBOX_COMMANDS,
  DROPBOX_PROMPT,
  DROPBOX_VFS_OPS,
  DropboxAccessor,
  DropboxTokenManager,
  PathSpec,
  RAMIndexCacheStore,
  ResourceName,
  dropboxRead,
  dropboxReaddir,
  dropboxResolveGlob,
  dropboxStat
} from "@struktoai/mirage-core";
var DropboxResource = class {
  kind = ResourceName.DROPBOX;
  isRemote = true;
  indexTtl = 86400;
  prompt = DROPBOX_PROMPT;
  config;
  accessor;
  index;
  constructor(config) {
    this.config = config;
    const tm = new DropboxTokenManager({
      clientId: config.clientId,
      clientSecret: config.clientSecret,
      refreshToken: config.refreshToken,
      ...config.refreshFn !== void 0 ? { refreshFn: config.refreshFn } : {}
    });
    this.accessor = new DropboxAccessor({ tokenManager: tm });
    this.index = new RAMIndexCacheStore({ ttl: 86400 });
  }
  open() {
    return Promise.resolve();
  }
  close() {
    return Promise.resolve();
  }
  commands() {
    return DROPBOX_COMMANDS;
  }
  ops() {
    return DROPBOX_VFS_OPS;
  }
  readFile(p) {
    return dropboxRead(this.accessor, p, this.index);
  }
  readdir(p) {
    return dropboxReaddir(this.accessor, p, this.index);
  }
  stat(p) {
    return dropboxStat(this.accessor, p, this.index);
  }
  async fingerprint(p) {
    const lookup = await this.index.get(p.original);
    return lookup.entry?.remoteTime ?? null;
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return dropboxResolveGlob(this.accessor, effective, this.index);
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["clientSecret", "refreshToken"],
      config: redactDropboxConfig(this.config)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  DropboxResource
};
