// @struktoai/mirage-node/dist/chunk-UWMOTOXA.js
import {
  redactGmailConfig
} from "./chunk-GO67AOYK.js";

// src/resource/gmail/gmail.ts
import {
  GMAIL_COMMANDS,
  GMAIL_PROMPT,
  GMAIL_WRITE_PROMPT,
  GMAIL_VFS_OPS,
  GmailAccessor,
  PathSpec,
  RAMIndexCacheStore,
  ResourceName,
  TokenManager,
  gmailRead,
  gmailReaddir,
  gmailResolveGlob,
  gmailStat
} from "@struktoai/mirage-core";
var GmailResource = class {
  kind = ResourceName.GMAIL;
  isRemote = true;
  indexTtl = 86400;
  prompt = GMAIL_PROMPT;
  writePrompt = GMAIL_WRITE_PROMPT;
  config;
  accessor;
  index;
  constructor(config) {
    this.config = config;
    const tm = new TokenManager({
      clientId: config.clientId,
      clientSecret: config.clientSecret,
      refreshToken: config.refreshToken
    });
    this.accessor = new GmailAccessor({ tokenManager: tm });
    this.index = new RAMIndexCacheStore({ ttl: 86400 });
  }
  open() {
    return Promise.resolve();
  }
  close() {
    return Promise.resolve();
  }
  commands() {
    return GMAIL_COMMANDS;
  }
  ops() {
    return GMAIL_VFS_OPS;
  }
  readFile(p) {
    return gmailRead(this.accessor, p, this.index);
  }
  readdir(p) {
    return gmailReaddir(this.accessor, p, this.index);
  }
  stat(p) {
    return gmailStat(this.accessor, p, this.index);
  }
  async fingerprint(p) {
    const lookup = await this.index.get(p.original);
    return lookup.entry?.remoteTime ?? null;
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return gmailResolveGlob(this.accessor, effective, this.index);
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["clientSecret", "refreshToken"],
      config: redactGmailConfig(this.config)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  GmailResource
};
