// @struktoai/mirage-node/dist/chunk-LPMPTQYP.js
// src/resource/s3/config.ts
import {
  normalizeFields
} from "@struktoai/mirage-core";
function redactConfig(config) {
  const { presignedUrlProvider, ...rest } = config;
  const out = { ...rest };
  if (config.accessKeyId !== void 0) out.accessKeyId = "<REDACTED>";
  if (config.secretAccessKey !== void 0) out.secretAccessKey = "<REDACTED>";
  if (config.sessionToken !== void 0) out.sessionToken = "<REDACTED>";
  if (presignedUrlProvider !== void 0) out.presignedUrlProvider = "<REDACTED>";
  return out;
}
function normalizeS3Config(input) {
  return normalizeFields(input, {
    rename: {
      aws_access_key_id: "accessKeyId",
      aws_secret_access_key: "secretAccessKey",
      aws_session_token: "sessionToken",
      aws_profile: "profile",
      endpoint_url: "endpoint",
      path_style: "forcePathStyle",
      timeout: "timeoutMs"
    },
    transform: {
      timeout: (v) => typeof v === "number" ? v * 1e3 : v
    },
    drop: ["proxy"]
  });
}

export {
  redactConfig,
  normalizeS3Config
};
