// @struktoai/mirage-node/dist/chunk-HXWGG4K7.js
import {
  S3Resource
} from "./chunk-G7O6DAWU.js";
import {
  ociToS3Config,
  redactOciConfig
} from "./chunk-ZC4KNEOQ.js";

// src/resource/oci/oci.ts
import { ResourceName } from "@struktoai/mirage-core";

// src/resource/oci/prompt.ts
var OCI_PROMPT = `{prefix}
  Remote Oracle Cloud Object Storage.
  IMPORTANT: This is a remote mount. Prefer targeted reads over full scans.
  Supports: ls, cat, head, tail, grep, rg, wc, find, tree, jq, stat.`;

// src/resource/oci/oci.ts
function remapOpsResource(ops, to) {
  return ops.map((op) => op.resource === ResourceName.S3 ? { ...op, resource: to } : op);
}
function remapCommandsResource(commands, to) {
  return commands.map((cmd) => {
    if (cmd.resource !== ResourceName.S3) return cmd;
    const proto = Object.getPrototypeOf(cmd);
    const copy = Object.assign(Object.create(proto), cmd);
    Object.assign(copy, { resource: to });
    return copy;
  });
}
var OCIResource = class extends S3Resource {
  kind = ResourceName.OCI;
  prompt = OCI_PROMPT;
  ociConfig;
  ociOps;
  ociCommands;
  constructor(config) {
    super(ociToS3Config(config));
    this.ociConfig = config;
    this.ociOps = remapOpsResource(super.ops(), ResourceName.OCI);
    this.ociCommands = remapCommandsResource(super.commands(), ResourceName.OCI);
  }
  ops() {
    return this.ociOps;
  }
  commands() {
    return this.ociCommands;
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["accessKeyId", "secretAccessKey"],
      config: redactOciConfig(this.ociConfig)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  OCI_PROMPT,
  OCIResource
};
