// @struktoai/mirage-node/dist/chunk-H74QYTN7.js
// src/resource/email/config.ts
import { normalizeFields } from "@struktoai/mirage-core";
function asString(v) {
  return typeof v === "string" ? v : "";
}
function redactEmailConfig(config) {
  return { ...config, password: "<REDACTED>" };
}
function buildEmailConfig(input) {
  return {
    imapHost: input.imapHost,
    imapPort: input.imapPort ?? 993,
    smtpHost: input.smtpHost,
    smtpPort: input.smtpPort ?? 587,
    username: input.username,
    password: input.password,
    useSsl: input.useSsl ?? true,
    maxMessages: input.maxMessages ?? 200
  };
}
function normalizeEmailConfig(input) {
  const norm = normalizeFields(input, {
    rename: {
      imap_host: "imapHost",
      imap_port: "imapPort",
      smtp_host: "smtpHost",
      smtp_port: "smtpPort",
      use_ssl: "useSsl",
      max_messages: "maxMessages"
    }
  });
  const built = {
    imapHost: asString(norm.imapHost),
    smtpHost: asString(norm.smtpHost),
    username: asString(norm.username),
    password: asString(norm.password)
  };
  if (typeof norm.imapPort === "number") built.imapPort = norm.imapPort;
  if (typeof norm.smtpPort === "number") built.smtpPort = norm.smtpPort;
  if (typeof norm.useSsl === "boolean") built.useSsl = norm.useSsl;
  if (typeof norm.maxMessages === "number") built.maxMessages = norm.maxMessages;
  return buildEmailConfig(built);
}

export {
  redactEmailConfig,
  buildEmailConfig,
  normalizeEmailConfig
};
