// @struktoai/mirage-node/dist/chunk-MVE5VVMQ.js
import {
  redactGDriveConfig
} from "./chunk-35PVZRWU.js";

// src/resource/gdrive/gdrive.ts
import {
  GDRIVE_COMMANDS,
  GDRIVE_PROMPT,
  GDRIVE_VFS_OPS,
  GDriveAccessor,
  PathSpec,
  RAMIndexCacheStore,
  ResourceName,
  TokenManager,
  gdriveRead,
  gdriveReaddir,
  gdriveResolveGlob,
  gdriveStat
} from "@struktoai/mirage-core";
var GDriveResource = class {
  kind = ResourceName.GDRIVE;
  isRemote = true;
  indexTtl = 86400;
  prompt = GDRIVE_PROMPT;
  config;
  accessor;
  index;
  constructor(config) {
    this.config = config;
    const tm = new TokenManager({
      clientId: config.clientId,
      clientSecret: config.clientSecret,
      refreshToken: config.refreshToken
    });
    this.accessor = new GDriveAccessor({ tokenManager: tm });
    this.index = new RAMIndexCacheStore({ ttl: 86400 });
  }
  open() {
    return Promise.resolve();
  }
  close() {
    return Promise.resolve();
  }
  commands() {
    return GDRIVE_COMMANDS;
  }
  ops() {
    return GDRIVE_VFS_OPS;
  }
  readFile(p) {
    return gdriveRead(this.accessor, p, this.index);
  }
  readdir(p) {
    return gdriveReaddir(this.accessor, p, this.index);
  }
  stat(p) {
    return gdriveStat(this.accessor, p, this.index);
  }
  async fingerprint(p) {
    const lookup = await this.index.get(p.original);
    return lookup.entry?.remoteTime ?? null;
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return gdriveResolveGlob(this.accessor, effective, this.index);
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["clientSecret", "refreshToken"],
      config: redactGDriveConfig(this.config)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  GDriveResource
};
