// @struktoai/mirage-node/dist/chunk-LBBM3LKK.js
import {
  redactTrelloConfig
} from "./chunk-QWWOBAYU.js";

// src/resource/trello/trello.ts
import {
  HttpTrelloTransport,
  PathSpec,
  RAMIndexCacheStore,
  ResourceName,
  resolveTrelloGlob,
  TRELLO_COMMANDS,
  TRELLO_PROMPT,
  TRELLO_VFS_OPS,
  TRELLO_WRITE_PROMPT,
  TrelloAccessor,
  trelloRead,
  trelloReaddir,
  trelloStat
} from "@struktoai/mirage-core";
var TrelloResource = class {
  kind = ResourceName.TRELLO;
  isRemote = true;
  indexTtl = 600;
  prompt = TRELLO_PROMPT;
  writePrompt = TRELLO_WRITE_PROMPT;
  config;
  accessor;
  index;
  constructor(config) {
    this.config = config;
    const transportOpts = {
      apiKey: config.apiKey,
      apiToken: config.apiToken
    };
    if (config.baseUrl !== void 0) transportOpts.baseUrl = config.baseUrl;
    this.accessor = new TrelloAccessor(new HttpTrelloTransport(transportOpts));
    this.index = new RAMIndexCacheStore({ ttl: this.indexTtl });
  }
  open() {
    return Promise.resolve();
  }
  close() {
    return Promise.resolve();
  }
  commands() {
    return TRELLO_COMMANDS;
  }
  ops() {
    return TRELLO_VFS_OPS;
  }
  filter() {
    const out = {};
    if (this.config.workspaceId !== void 0) out.workspaceId = this.config.workspaceId;
    if (this.config.boardIds !== void 0) out.boardIds = this.config.boardIds;
    return out;
  }
  readFile(p) {
    return trelloRead(this.accessor, p, this.index);
  }
  readdir(p) {
    return trelloReaddir(this.accessor, p, this.index, this.filter());
  }
  stat(p) {
    return trelloStat(this.accessor, p, this.index);
  }
  async fingerprint(p) {
    const lookup = await this.index.get(p.original);
    return lookup.entry?.remoteTime ?? null;
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return resolveTrelloGlob(this.accessor, effective, this.index, this.filter());
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["apiKey", "apiToken"],
      config: redactTrelloConfig(this.config)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  TrelloResource
};
