// @struktoai/mirage-node/dist/chunk-PEZYBKZH.js
import {
  redactGitHubConfig
} from "./chunk-VCLMTMSO.js";

// src/resource/github/github.ts
import {
  fetchGitHubRepoInfo,
  fetchGitHubTree,
  GITHUB_COMMANDS,
  GITHUB_VFS_OPS,
  GitHubAccessor,
  HttpGitHubTransport,
  PathSpec,
  RAMIndexCacheStore,
  ResourceName,
  githubIndexEntryFromTree,
  githubMakeTreeEntry,
  githubRead,
  githubReaddir,
  githubResolveGlob,
  githubStat
} from "@struktoai/mirage-core";
var GitHubResource = class _GitHubResource {
  kind = ResourceName.GITHUB;
  isRemote = true;
  indexTtl = 86400;
  config;
  accessor;
  index;
  constructor(config, accessor, index) {
    this.config = config;
    this.accessor = accessor;
    this.index = index;
  }
  static async create(config) {
    const transportOpts = { token: config.token };
    if (config.baseUrl !== void 0) transportOpts.baseUrl = config.baseUrl;
    const transport = new HttpGitHubTransport(transportOpts);
    const repoInfo = await fetchGitHubRepoInfo(transport, config.owner, config.repo);
    const ref = config.ref ?? repoInfo.default_branch;
    const { tree, truncated } = await fetchGitHubTree(transport, config.owner, config.repo, ref);
    const treeMap = {};
    for (const item of tree) treeMap[item.path] = githubMakeTreeEntry(item);
    const accessor = new GitHubAccessor({
      transport,
      owner: config.owner,
      repo: config.repo,
      ref,
      defaultBranch: repoInfo.default_branch,
      truncated,
      tree: treeMap
    });
    const index = new RAMIndexCacheStore({ ttl: 86400 });
    await populateIndex(index, tree);
    return new _GitHubResource(config, accessor, index);
  }
  open() {
    return Promise.resolve();
  }
  close() {
    return Promise.resolve();
  }
  commands() {
    return GITHUB_COMMANDS;
  }
  ops() {
    return GITHUB_VFS_OPS;
  }
  readFile(p) {
    return githubRead(this.accessor, p, this.index);
  }
  readdir(p) {
    return githubReaddir(this.accessor, p, this.index);
  }
  stat(p) {
    return githubStat(this.accessor, p, this.index);
  }
  async fingerprint(p) {
    const lookup = await this.index.get(p.original);
    return lookup.entry?.id ?? null;
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return githubResolveGlob(this.accessor, effective, this.index);
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["token"],
      config: redactGitHubConfig(this.config),
      defaultBranch: this.accessor.defaultBranch,
      truncated: this.accessor.truncated
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};
async function populateIndex(index, tree) {
  const dirs = /* @__PURE__ */ new Map();
  for (const item of tree) {
    const parts = item.path.split("/");
    const name = parts[parts.length - 1] ?? item.path;
    const parent = parts.length > 1 ? `/${parts.slice(0, -1).join("/")}` : "/";
    const arr = dirs.get(parent) ?? [];
    arr.push([name, githubIndexEntryFromTree(item)]);
    dirs.set(parent, arr);
  }
  await Promise.all([...dirs].map(([parent, entries]) => index.setDir(parent, entries)));
}

export {
  GitHubResource
};
