// @struktoai/mirage-node/dist/chunk-GCB4JUF7.js
import {
  redactGitHubCIConfig
} from "./chunk-GPF66TH2.js";

// src/resource/github_ci/github_ci.ts
import {
  GITHUB_CI_COMMANDS,
  GITHUB_CI_PROMPT,
  GITHUB_CI_VFS_OPS,
  GitHubCIAccessor,
  HttpCITransport,
  PathSpec,
  RAMIndexCacheStore,
  ResourceName,
  githubCiRead,
  githubCiReaddir,
  githubCiResolveGlob,
  githubCiStat
} from "@struktoai/mirage-core";
var GitHubCIResource = class {
  kind = ResourceName.GITHUB_CI;
  isRemote = true;
  indexTtl = 86400;
  prompt = GITHUB_CI_PROMPT;
  config;
  accessor;
  index;
  constructor(config) {
    this.config = config;
    const transportOpts = { token: config.token };
    if (config.baseUrl !== void 0) transportOpts.baseUrl = config.baseUrl;
    this.accessor = new GitHubCIAccessor({
      transport: new HttpCITransport(transportOpts),
      owner: config.owner,
      repo: config.repo,
      days: config.days ?? 30,
      maxRuns: config.maxRuns ?? 300
    });
    this.index = new RAMIndexCacheStore({ ttl: 86400 });
  }
  open() {
    return Promise.resolve();
  }
  close() {
    return Promise.resolve();
  }
  commands() {
    return GITHUB_CI_COMMANDS;
  }
  ops() {
    return GITHUB_CI_VFS_OPS;
  }
  readFile(p) {
    return githubCiRead(this.accessor, p, this.index);
  }
  readdir(p) {
    return githubCiReaddir(this.accessor, p, this.index);
  }
  stat(p) {
    return githubCiStat(this.accessor, p, this.index);
  }
  async fingerprint(p) {
    const lookup = await this.index.get(p.original);
    return lookup.entry?.remoteTime ?? null;
  }
  glob(paths, prefix = "") {
    const effective = prefix !== "" ? paths.map(
      (p) => p.prefix !== "" ? p : new PathSpec({
        original: p.original,
        directory: p.directory,
        ...p.pattern !== null ? { pattern: p.pattern } : {},
        resolved: p.resolved,
        prefix
      })
    ) : paths;
    return githubCiResolveGlob(this.accessor, effective, this.index);
  }
  getState() {
    return Promise.resolve({
      type: this.kind,
      needsOverride: true,
      redactedFields: ["token"],
      config: redactGitHubCIConfig(this.config)
    });
  }
  loadState(_state) {
    return Promise.resolve();
  }
};

export {
  GitHubCIResource
};
