// @struktoai/mirage-core/dist/_client_browser-KPKXNOAW.js
// src/core/s3/_client_browser.ts
function decodeEntities(s) {
  return s.replaceAll("&amp;", "&").replaceAll("&lt;", "<").replaceAll("&gt;", ">").replaceAll("&quot;", '"').replaceAll("&apos;", "'");
}
function tagText(xml, tag) {
  const re = new RegExp(`<${tag}[^>]*>([\\s\\S]*?)</${tag}>`);
  const m = re.exec(xml);
  return m === null ? null : decodeEntities(m[1] ?? "");
}
function tagBlocks(xml, tag) {
  const re = new RegExp(`<${tag}[^>]*>([\\s\\S]*?)</${tag}>`, "g");
  const blocks = [];
  let m;
  while ((m = re.exec(xml)) !== null) blocks.push(m[1] ?? "");
  return blocks;
}
function parseListObjectsV2(xml) {
  const body = tagText(xml, "ListBucketResult") ?? xml;
  const truncated = (tagText(body, "IsTruncated") ?? "false").toLowerCase() === "true";
  const nextToken = tagText(body, "NextContinuationToken") ?? void 0;
  const prefixes = [];
  for (const blk of tagBlocks(body, "CommonPrefixes")) {
    const p = tagText(blk, "Prefix");
    if (p !== null && p !== "") prefixes.push({ Prefix: p });
  }
  const contents = [];
  for (const blk of tagBlocks(body, "Contents")) {
    const key = tagText(blk, "Key");
    if (key === null) continue;
    const sizeStr = tagText(blk, "Size");
    const size = sizeStr !== null ? Number.parseInt(sizeStr, 10) : 0;
    const lastMod = tagText(blk, "LastModified");
    const etagRaw = tagText(blk, "ETag");
    const etag = etagRaw !== null ? etagRaw.replaceAll('"', "") : "";
    const entry = {
      Key: key,
      Size: Number.isFinite(size) ? size : 0
    };
    if (lastMod !== null) entry.LastModified = new Date(lastMod);
    if (etag !== "") entry.ETag = etag;
    contents.push(entry);
  }
  const out = {
    CommonPrefixes: prefixes,
    Contents: contents,
    IsTruncated: truncated
  };
  if (nextToken !== void 0) out.NextContinuationToken = nextToken;
  return out;
}
var BrowserCommand = class {
  __mirageTag;
  input;
  constructor(tag, input) {
    this.__mirageTag = tag;
    this.input = input;
  }
};
var GetObjectCommand = class extends BrowserCommand {
  constructor(input) {
    super("Get", input);
  }
};
var HeadObjectCommand = class extends BrowserCommand {
  constructor(input) {
    super("Head", input);
  }
};
var ListObjectsV2Command = class extends BrowserCommand {
  constructor(input) {
    super("List", input);
  }
};
var PutObjectCommand = class extends BrowserCommand {
  constructor(input) {
    super("Put", input);
  }
};
var DeleteObjectCommand = class extends BrowserCommand {
  constructor(input) {
    super("Delete", input);
  }
};
var DeleteObjectsCommand = class extends BrowserCommand {
  constructor(input) {
    super("Delete", input);
  }
};
var CopyObjectCommand = class extends BrowserCommand {
  constructor(input) {
    super("Copy", input);
  }
};
var stubS3ClientCtor = function stubS3Client() {
  throw new Error(
    "BROWSER_S3_MODULE.S3Client is a stub \u2014 use createS3Client(config) to get the presigned-fetch client"
  );
};
var BROWSER_S3_MODULE = {
  S3Client: stubS3ClientCtor,
  GetObjectCommand,
  HeadObjectCommand,
  ListObjectsV2Command,
  PutObjectCommand,
  DeleteObjectCommand,
  DeleteObjectsCommand,
  CopyObjectCommand
};
async function bodyFromResponse(resp) {
  const stream = resp.body;
  if (stream === null) {
    const buf = new Uint8Array(await resp.arrayBuffer());
    return (async function* () {
      await Promise.resolve();
      yield buf;
    })();
  }
  const reader = stream.getReader();
  return (async function* () {
    try {
      for (; ; ) {
        const { value, done } = await reader.read();
        if (done) return;
        yield value;
      }
    } finally {
      reader.releaseLock();
    }
  })();
}
function keyFromInput(input) {
  const k = input.Key;
  return typeof k === "string" ? k : "";
}
function makeNotFound(path) {
  const e = new Error(`S3 object not found: ${path}`);
  e.$metadata = { httpStatusCode: 404 };
  return e;
}
async function sendBrowserCommand(provider, defaultContentType, cmd) {
  const browserCmd = cmd;
  if (typeof browserCmd.__mirageTag !== "string") {
    throw new Error(
      "presigned-fetch client received a non-browser S3 command; check that you did not mix AWS SDK command classes with a presigner-backed config"
    );
  }
  const tag = browserCmd.__mirageTag;
  switch (tag) {
    case "Get": {
      const input = browserCmd.input;
      const key = typeof input.Key === "string" ? input.Key : "";
      const url = await provider(`/${key}`, "GET");
      const resp = await fetch(url);
      if (resp.status === 404) throw makeNotFound(key);
      if (!resp.ok) {
        throw new Error(`S3 GET ${key} failed: ${String(resp.status)} ${resp.statusText}`);
      }
      const body = await bodyFromResponse(resp);
      const lenHeader = resp.headers.get("content-length");
      const ContentLength = lenHeader !== null ? Number.parseInt(lenHeader, 10) : void 0;
      const out = { Body: body };
      if (ContentLength !== void 0 && !Number.isNaN(ContentLength)) {
        out.ContentLength = ContentLength;
      }
      const etag = resp.headers.get("etag")?.replace(/^"|"$/g, "") ?? "";
      if (etag !== "") out.ETag = etag;
      const lm = resp.headers.get("last-modified");
      if (lm !== null) out.LastModified = new Date(lm);
      return out;
    }
    case "Head": {
      const key = keyFromInput(browserCmd.input);
      const url = await provider(`/${key}`, "HEAD");
      const resp = await fetch(url, { method: "HEAD" });
      if (resp.status === 404) throw makeNotFound(key);
      if (!resp.ok) {
        throw new Error(`S3 HEAD ${key} failed: ${String(resp.status)} ${resp.statusText}`);
      }
      const lenHeader = resp.headers.get("content-length");
      const ContentLength = lenHeader !== null ? Number.parseInt(lenHeader, 10) : void 0;
      const etag = resp.headers.get("etag")?.replace(/^"|"$/g, "") ?? "";
      const lm = resp.headers.get("last-modified");
      const out = {};
      if (ContentLength !== void 0 && !Number.isNaN(ContentLength)) {
        out.ContentLength = ContentLength;
      }
      if (etag !== "") out.ETag = etag;
      if (lm !== null) out.LastModified = new Date(lm);
      return out;
    }
    case "List": {
      const input = browserCmd.input;
      const opts = {
        listPrefix: input.Prefix ?? ""
      };
      if (input.Delimiter !== void 0) opts.listDelimiter = input.Delimiter;
      if (input.ContinuationToken !== void 0)
        opts.listContinuationToken = input.ContinuationToken;
      const url = await provider("/", "LIST", opts);
      const resp = await fetch(url);
      if (!resp.ok) {
        throw new Error(`S3 LIST failed: ${String(resp.status)} ${resp.statusText}`);
      }
      const xml = await resp.text();
      return parseListObjectsV2(xml);
    }
    case "Put": {
      const input = browserCmd.input;
      const key = typeof input.Key === "string" ? input.Key : "";
      const contentType = typeof input.ContentType === "string" ? input.ContentType : defaultContentType ?? "application/octet-stream";
      const url = await provider(`/${key}`, "PUT", { contentType });
      const body = input.Body;
      const resp = await fetch(url, {
        method: "PUT",
        headers: { "Content-Type": contentType },
        body
      });
      if (!resp.ok) {
        const errBody = await resp.text().catch(() => "");
        throw new Error(
          `S3 PUT ${key} failed: ${String(resp.status)} ${resp.statusText}${errBody !== "" ? ` \u2014 ${errBody.slice(0, 200)}` : ""}`
        );
      }
      return {};
    }
    case "Delete": {
      const input = browserCmd.input;
      if (typeof input.Key === "string") {
        const url = await provider(`/${input.Key}`, "DELETE");
        const resp = await fetch(url, { method: "DELETE" });
        if (resp.status !== 404 && !resp.ok) {
          throw new Error(
            `S3 DELETE ${input.Key} failed: ${String(resp.status)} ${resp.statusText}`
          );
        }
        return {};
      }
      const objects = input.Delete?.Objects ?? [];
      await Promise.all(
        objects.map(async (o) => {
          const url = await provider(`/${o.Key}`, "DELETE");
          await fetch(url, { method: "DELETE" });
        })
      );
      return { Deleted: objects.map((o) => ({ Key: o.Key })) };
    }
    case "Copy": {
      const input = browserCmd.input;
      const dstKey = typeof input.Key === "string" ? input.Key : "";
      const copySrcRaw = typeof input.CopySource === "string" ? input.CopySource : "";
      const srcKey = copySrcRaw.replace(/^\/?[^/]+\//, "");
      const url = await provider(`/${dstKey}`, "COPY", { copySource: srcKey });
      const resp = await fetch(url, {
        method: "PUT",
        headers: { "x-amz-copy-source": `/${srcKey}` }
      });
      if (!resp.ok) {
        const errBody = await resp.text().catch(() => "");
        throw new Error(
          `S3 COPY ${copySrcRaw} \u2192 ${dstKey} failed: ${String(resp.status)} ${resp.statusText}${errBody !== "" ? ` \u2014 ${errBody.slice(0, 200)}` : ""}`
        );
      }
      return {};
    }
  }
}
function createBrowserS3Client(config) {
  const provider = config.presignedUrlProvider;
  if (provider === void 0) {
    throw new Error("createBrowserS3Client called without presignedUrlProvider");
  }
  return {
    send: (cmd) => sendBrowserCommand(provider, config.defaultContentType, cmd)
  };
}
export {
  BROWSER_S3_MODULE,
  createBrowserS3Client
};
