// @struktoai/mirage-core/dist/chunk-CYEPQFOB.js
// src/utils/optional_peer.ts
async function loadOptionalPeer(importer, config) {
  try {
    return await importer();
  } catch (err) {
    const code = err?.code;
    if (code !== "ERR_MODULE_NOT_FOUND" && code !== "MODULE_NOT_FOUND") throw err;
    const docsLine = config.docsUrl !== void 0 ? `
See ${config.docsUrl} for details.` : "";
    throw new Error(
      `${config.feature} requires the optional peer dependency \`${config.packageName}\`. Install it with:

    pnpm add ${config.packageName}
` + docsLine
    );
  }
}

// src/core/s3/_client.ts
function s3Key(path) {
  return path.replace(/^\/+/, "");
}
function s3Prefix(path) {
  const key = s3Key(path);
  if (key === "") return "";
  return key.endsWith("/") ? key : `${key}/`;
}
function rawPathOf(path) {
  const prefix = path.prefix;
  return prefix !== "" && path.original.startsWith(prefix) ? path.original.slice(prefix.length) || "/" : path.original;
}
function fnmatch(name, pattern) {
  let re = "^";
  for (const ch of pattern) {
    if (ch === "*") re += ".*";
    else if (ch === "?") re += ".";
    else if (/[.+^${}()|[\]\\]/.test(ch)) re += "\\" + ch;
    else re += ch;
  }
  re += "$";
  return new RegExp(re).test(name);
}
async function withClient(config, fn) {
  const client = await createS3Client(config);
  try {
    return await fn(client);
  } finally {
    client.destroy?.();
  }
}
var cachedModule = null;
async function loadS3Module(config) {
  if (config?.presignedUrlProvider !== void 0) {
    const { BROWSER_S3_MODULE } = await import("./_client_browser-KPKXNOAW.js");
    return BROWSER_S3_MODULE;
  }
  cachedModule ??= loadOptionalPeer(
    () => import("@aws-sdk/client-s3"),
    {
      feature: "S3Resource",
      packageName: "@aws-sdk/client-s3",
      docsUrl: "https://mirage.dev/typescript/install"
    }
  );
  return cachedModule;
}
async function createS3Client(config) {
  if (config.presignedUrlProvider !== void 0) {
    const { createBrowserS3Client } = await import("./_client_browser-KPKXNOAW.js");
    return createBrowserS3Client(config);
  }
  const mod = await loadS3Module();
  const options = {};
  if (config.region !== void 0) options.region = config.region;
  if (config.endpoint !== void 0) options.endpoint = config.endpoint;
  if (config.forcePathStyle === true) options.forcePathStyle = true;
  if (config.accessKeyId !== void 0 && config.secretAccessKey !== void 0) {
    options.credentials = {
      accessKeyId: config.accessKeyId,
      secretAccessKey: config.secretAccessKey,
      ...config.sessionToken !== void 0 ? { sessionToken: config.sessionToken } : {}
    };
  }
  if (config.timeoutMs !== void 0) {
    options.requestHandler = {
      connectionTimeout: config.timeoutMs,
      requestTimeout: config.timeoutMs
    };
  }
  return new mod.S3Client(options);
}
function isNotFoundError(err) {
  if (err === null || typeof err !== "object") return false;
  const e = err;
  if (e.name === "NoSuchKey" || e.name === "NotFound") return true;
  if (e.Code === "NoSuchKey" || e.Code === "404") return true;
  return e.$metadata?.httpStatusCode === 404;
}
async function streamToBuffer(stream) {
  if (stream === null || stream === void 0) return new Uint8Array();
  const s = stream;
  if (typeof s.transformToByteArray === "function") {
    return s.transformToByteArray();
  }
  if (typeof s[Symbol.asyncIterator] === "function") {
    const chunks = [];
    let total = 0;
    for await (const chunk of s) {
      chunks.push(chunk);
      total += chunk.byteLength;
    }
    const out = new Uint8Array(total);
    let offset = 0;
    for (const c of chunks) {
      out.set(c, offset);
      offset += c.byteLength;
    }
    return out;
  }
  throw new Error("S3 response Body is not a readable stream");
}

// src/core/s3/read.ts
async function read(accessor, path, _index, options = {}) {
  const original = path.original;
  const prefix = path.prefix;
  const rawPath = prefix !== "" && original.startsWith(prefix) ? original.slice(prefix.length) || "/" : original;
  const key = s3Key(rawPath);
  const { config } = accessor;
  const { GetObjectCommand } = await loadS3Module(config);
  const client = await createS3Client(config);
  const input = { Bucket: config.bucket, Key: key };
  if (options.offset !== void 0 || options.size !== void 0) {
    const start = options.offset ?? 0;
    const end = options.size !== void 0 ? start + options.size - 1 : "";
    input.Range = `bytes=${String(start)}-${String(end)}`;
  }
  try {
    const resp = await client.send(new GetObjectCommand(input));
    return await streamToBuffer(resp.Body);
  } catch (err) {
    if (isNotFoundError(err)) {
      const e = new Error(`S3 object not found: ${rawPath}`);
      e.code = "ENOENT";
      throw e;
    }
    throw err;
  } finally {
    ;
    client.destroy?.();
  }
}

export {
  loadOptionalPeer,
  s3Key,
  s3Prefix,
  rawPathOf,
  fnmatch,
  withClient,
  loadS3Module,
  createS3Client,
  isNotFoundError,
  streamToBuffer,
  read
};
