// @struktoai/mirage-core/dist/chunk-E7N6O2ON.js
// src/core/jq/eval.ts
import * as jqWasm from "jq-wasm";

// src/core/jq/format.ts
var JQ_EMPTY = /* @__PURE__ */ Symbol("JQ_EMPTY");
var ENC = new TextEncoder();
function formatOne(value, raw2, compact) {
  if (raw2 && typeof value === "string") return ENC.encode(value + "\n");
  const json = compact ? JSON.stringify(value) : JSON.stringify(value, null, 2);
  return ENC.encode(json + "\n");
}
function concatBytes(parts) {
  let total = 0;
  for (const p of parts) total += p.byteLength;
  const out = new Uint8Array(total);
  let offset = 0;
  for (const p of parts) {
    out.set(p, offset);
    offset += p.byteLength;
  }
  return out;
}
function formatJqOutput(result, raw2, compact, spread) {
  if (result === JQ_EMPTY) return new Uint8Array(0);
  if (spread && Array.isArray(result)) {
    const parts = [];
    for (const item of result) parts.push(formatOne(item, raw2, compact));
    return concatBytes(parts);
  }
  return formatOne(result, raw2, compact);
}

// src/core/jq/eval.ts
function hasTopLevelSpread(expr) {
  let depth = 0;
  let inStr = false;
  for (let i = 0; i < expr.length; i++) {
    const ch = expr[i];
    if (ch === '"' && (i === 0 || expr[i - 1] !== "\\")) {
      inStr = !inStr;
      continue;
    }
    if (inStr) continue;
    if (depth === 0 && ch === "[" && expr[i + 1] === "]") return true;
    if (ch === "(" || ch === "[" || ch === "{") depth += 1;
    else if (ch === ")" || ch === "]" || ch === "}") depth -= 1;
  }
  return false;
}
async function jqEval(obj, expr) {
  const result = await jqWasm.raw(JSON.stringify(obj), expr, ["-c"]);
  if (result.exitCode !== 0) {
    throw new Error(result.stderr || `jq exited with code ${String(result.exitCode)}`);
  }
  const lines = result.stdout === "" ? [] : result.stdout.split("\n").filter((l) => l !== "");
  const outputs = lines.map((l) => JSON.parse(l));
  if (outputs.length === 0) return JQ_EMPTY;
  if (outputs.length === 1 && !hasTopLevelSpread(expr)) return outputs[0];
  return outputs;
}

export {
  JQ_EMPTY,
  concatBytes,
  formatJqOutput,
  jqEval
};
