// @struktoai/mirage-core/dist/chunk-OJUKX27E.js
// src/io/cachable_iterator.ts
var CachableAsyncIterator = class {
  source;
  buffer = [];
  exhaustedFlag = false;
  drainResolve = null;
  drainPromise;
  constructor(source) {
    this.source = source[Symbol.asyncIterator]();
    this.drainPromise = new Promise((resolve) => {
      this.drainResolve = resolve;
    });
  }
  get exhausted() {
    return this.exhaustedFlag;
  }
  [Symbol.asyncIterator]() {
    return this;
  }
  async next() {
    try {
      const result = await this.source.next();
      if (result.done === true) {
        this.markExhausted();
        return { done: true, value: void 0 };
      }
      this.buffer.push(result.value);
      return { done: false, value: result.value };
    } catch (err) {
      this.markExhausted();
      throw err;
    }
  }
  async drain() {
    try {
      for (; ; ) {
        const result = await this.source.next();
        if (result.done === true) break;
        this.buffer.push(result.value);
      }
    } finally {
      this.markExhausted();
    }
    return concat(this.buffer);
  }
  async drainBounded(maxBytes) {
    let total = 0;
    for (const c of this.buffer) total += c.byteLength;
    try {
      for (; ; ) {
        const result = await this.source.next();
        if (result.done === true) break;
        this.buffer.push(result.value);
        total += result.value.byteLength;
        if (total > maxBytes) return [concat(this.buffer), false];
      }
    } finally {
      this.markExhausted();
    }
    return [concat(this.buffer), true];
  }
  async waitForDrain() {
    await this.drainPromise;
    return concat(this.buffer);
  }
  markExhausted() {
    this.exhaustedFlag = true;
    if (this.drainResolve !== null) {
      const resolve = this.drainResolve;
      this.drainResolve = null;
      resolve();
    }
  }
};
function concat(chunks) {
  let total = 0;
  for (const c of chunks) total += c.byteLength;
  const out = new Uint8Array(total);
  let offset = 0;
  for (const c of chunks) {
    out.set(c, offset);
    offset += c.byteLength;
  }
  return out;
}

// src/io/types.ts
async function materialize(source) {
  if (source === null || source === void 0) return new Uint8Array();
  if (source instanceof Uint8Array) return source;
  if (source instanceof CachableAsyncIterator) return source.drain();
  const chunks = [];
  for await (const chunk of source) chunks.push(chunk);
  return concat2(chunks);
}
var IOResult = class _IOResult {
  stdout;
  stderr;
  exitCode;
  reads;
  writes;
  cache;
  streamSource;
  constructor(init = {}) {
    this.stdout = init.stdout ?? null;
    this.stderr = init.stderr ?? null;
    this.exitCode = init.exitCode ?? 0;
    this.reads = init.reads ?? {};
    this.writes = init.writes ?? {};
    this.cache = init.cache ?? [];
    this.streamSource = null;
  }
  async materializeStdout() {
    const bytes = await materialize(this.stdout);
    this.stdout = bytes;
    return bytes;
  }
  async stdoutStr(errors = "replace") {
    return decodeBytes(await this.materializeStdout(), errors);
  }
  async materializeStderr() {
    const bytes = await materialize(this.stderr);
    this.stderr = bytes;
    return bytes;
  }
  async stderrStr(errors = "replace") {
    return decodeBytes(await this.materializeStderr(), errors);
  }
  syncExitCode() {
    if (this.streamSource !== null) {
      this.streamSource.syncExitCode();
      this.exitCode = this.streamSource.exitCode;
    }
  }
  async merge(other) {
    const leftStderr = await materialize(this.stderr);
    const rightStderr = await materialize(other.stderr);
    let mergedStderr = null;
    if (leftStderr.byteLength > 0 || rightStderr.byteLength > 0) {
      mergedStderr = concat2([leftStderr, rightStderr]);
    }
    other.syncExitCode();
    const result = new _IOResult({
      stdout: other.stdout,
      stderr: mergedStderr,
      exitCode: other.exitCode,
      reads: { ...this.reads, ...other.reads },
      writes: { ...this.writes, ...other.writes },
      cache: [...this.cache, ...other.cache]
    });
    result.streamSource = other;
    return result;
  }
  async mergeAggregate(other) {
    const result = await this.merge(other);
    result.exitCode = Math.max(this.exitCode, other.exitCode);
    return result;
  }
};
function decodeBytes(bytes, errors) {
  return new TextDecoder("utf-8", { fatal: errors === "strict" }).decode(bytes);
}
function concat2(chunks) {
  let total = 0;
  for (const c of chunks) total += c.byteLength;
  const out = new Uint8Array(total);
  let offset = 0;
  for (const c of chunks) {
    out.set(c, offset);
    offset += c.byteLength;
  }
  return out;
}

// src/types.ts
var MountMode = Object.freeze({
  READ: "read",
  WRITE: "write",
  EXEC: "exec"
});
var ConsistencyPolicy = Object.freeze({
  LAZY: "lazy",
  ALWAYS: "always"
});
var ResourceName = Object.freeze({
  DISK: "disk",
  S3: "s3",
  RAM: "ram",
  GITHUB: "github",
  LINEAR: "linear",
  GDOCS: "gdocs",
  GSHEETS: "gsheets",
  GSLIDES: "gslides",
  GDRIVE: "gdrive",
  DROPBOX: "dropbox",
  BOX: "box",
  SLACK: "slack",
  DISCORD: "discord",
  GMAIL: "gmail",
  TRELLO: "trello",
  TELEGRAM: "telegram",
  MONGODB: "mongodb",
  NOTION: "notion",
  LANGFUSE: "langfuse",
  SSH: "ssh",
  REDIS: "redis",
  GITHUB_CI: "github_ci",
  GCS: "gcs",
  OCI: "oci",
  R2: "r2",
  EMAIL: "email",
  PAPERCLIP: "paperclip",
  OPFS: "opfs",
  SUPABASE: "supabase",
  POSTGRES: "postgres",
  SSCHOLAR_PAPER: "sscholar-paper",
  SSCHOLAR_AUTHOR: "sscholar-author",
  VERCEL: "vercel",
  POSTHOG: "posthog"
});
var DEFAULT_SESSION_ID = "default";
var DEFAULT_AGENT_ID = "default";
var FileType = Object.freeze({
  DIRECTORY: "directory",
  TEXT: "text",
  BINARY: "binary",
  JSON: "json",
  CSV: "csv",
  IMAGE_PNG: "image/png",
  IMAGE_JPEG: "image/jpeg",
  IMAGE_GIF: "image/gif",
  ZIP: "application/zip",
  GZIP: "application/gzip",
  PDF: "application/pdf",
  PARQUET: "parquet",
  ORC: "orc",
  FEATHER: "feather",
  HDF5: "hdf5"
});
var FileStat = class {
  name;
  size;
  modified;
  fingerprint;
  type;
  extra;
  constructor(init) {
    this.name = init.name;
    this.size = init.size ?? null;
    this.modified = init.modified ?? null;
    this.fingerprint = init.fingerprint ?? null;
    this.type = init.type ?? null;
    this.extra = init.extra ?? {};
    Object.freeze(this);
  }
};
var PathSpec = class _PathSpec {
  original;
  directory;
  pattern;
  resolved;
  prefix;
  constructor(init) {
    this.original = init.original;
    this.directory = init.directory;
    this.pattern = init.pattern ?? null;
    this.resolved = init.resolved ?? true;
    this.prefix = init.prefix ?? "";
    Object.freeze(this);
  }
  get stripPrefix() {
    if (this.prefix && this.original.startsWith(this.prefix)) {
      return this.original.slice(this.prefix.length) || "/";
    }
    return this.original;
  }
  get key() {
    return this.stripPrefix.replace(/^\/+|\/+$/g, "");
  }
  get dir() {
    return new _PathSpec({
      original: this.directory,
      directory: this.directory,
      pattern: this.pattern,
      resolved: false,
      prefix: this.prefix
    });
  }
  child(name) {
    return `${this.original.replace(/\/+$/, "")}/${name}`;
  }
  static fromStrPath(path, prefix = "") {
    const idx = path.lastIndexOf("/");
    const directory = path.slice(0, idx + 1) || "/";
    return new _PathSpec({ original: path, directory, prefix });
  }
};

// src/workspace/types.ts
var ExecutionNode = class {
  command;
  op;
  stderr;
  exitCode;
  children;
  records;
  constructor(init = {}) {
    this.command = init.command ?? null;
    this.op = init.op ?? null;
    this.stderr = init.stderr ?? new Uint8Array();
    this.exitCode = init.exitCode ?? 0;
    this.children = init.children ?? [];
    this.records = init.records ?? [];
  }
  toJSON() {
    const d = {};
    if (this.command !== null) d.command = this.command;
    if (this.op !== null) d.op = this.op;
    d.stderr = new TextDecoder("utf-8", { fatal: false }).decode(this.stderr);
    d.exitCode = this.exitCode;
    if (this.children.length > 0) d.children = this.children.map((c) => c.toJSON());
    if (this.records.length > 0) d.records = this.records.map((r) => r.toJSON());
    return d;
  }
};
var ExecutionRecord = class {
  agent;
  command;
  stdout;
  stdin;
  exitCode;
  tree;
  timestamp;
  sessionId;
  constructor(init) {
    this.agent = init.agent;
    this.command = init.command;
    this.stdout = init.stdout;
    this.stdin = init.stdin ?? null;
    this.exitCode = init.exitCode;
    this.tree = init.tree;
    this.timestamp = init.timestamp;
    this.sessionId = init.sessionId ?? DEFAULT_SESSION_ID;
  }
  toJSON() {
    const decoder = new TextDecoder("utf-8", { fatal: false });
    return {
      agent: this.agent,
      command: this.command,
      stdout: decoder.decode(this.stdout),
      stdin: this.stdin === null ? null : decoder.decode(this.stdin),
      exitCode: this.exitCode,
      tree: this.tree.toJSON(),
      timestamp: this.timestamp,
      sessionId: this.sessionId
    };
  }
};

// src/workspace/executor/python/types.ts
var PyodideUnavailableError = class extends Error {
  constructor(message, options) {
    super(message, options);
    this.name = "PyodideUnavailableError";
  }
};

// src/workspace/executor/python/handle.ts
function readAllBytes(data) {
  if (data instanceof Uint8Array) return Promise.resolve(data);
  if (data === null || data === void 0) return Promise.resolve(new Uint8Array());
  return materialize(data);
}
function toPathSpec(p) {
  return new PathSpec({
    original: p.original,
    directory: p.directory,
    pattern: p.pattern,
    resolved: p.resolved,
    prefix: p.prefix
  });
}
async function handlePython(dispatch, pathScope, args, opts, deps) {
  let code = opts.code;
  const cmdStr = pathScope !== null ? `python3 ${pathScope.original}` : "python3 -c";
  if (code === null) {
    if (pathScope === null) {
      const err = new TextEncoder().encode("python3: no input\n");
      return [
        null,
        new IOResult({ exitCode: 1, stderr: err }),
        new ExecutionNode({ command: cmdStr, exitCode: 1 })
      ];
    }
    try {
      const [data] = await dispatch("read", toPathSpec(pathScope));
      const bytes = await readAllBytes(data);
      code = new TextDecoder("utf-8", { fatal: false }).decode(bytes);
    } catch {
      const err = new TextEncoder().encode(`python3: ${pathScope.original}: No such file
`);
      return [
        null,
        new IOResult({ exitCode: 1, stderr: err }),
        new ExecutionNode({ command: cmdStr, exitCode: 1 })
      ];
    }
  }
  let stdinBytes = null;
  if (opts.stdin !== null) {
    stdinBytes = await materialize(opts.stdin);
  }
  try {
    const result = await deps.runtime.run({
      code,
      args,
      env: opts.env,
      stdin: stdinBytes
    });
    return [
      result.stdout.length > 0 ? result.stdout : null,
      new IOResult({ exitCode: result.exitCode, stderr: result.stderr }),
      new ExecutionNode({ command: cmdStr, exitCode: result.exitCode })
    ];
  } catch (err) {
    if (err instanceof PyodideUnavailableError) {
      return [
        null,
        new IOResult({
          exitCode: 127,
          stderr: new TextEncoder().encode(`python3: ${err.message}
`)
        }),
        new ExecutionNode({ command: cmdStr, exitCode: 127 })
      ];
    }
    const msg = err instanceof Error ? err.message : String(err);
    return [
      null,
      new IOResult({
        exitCode: 1,
        stderr: new TextEncoder().encode(`python3: ${msg}
`)
      }),
      new ExecutionNode({ command: cmdStr, exitCode: 1 })
    ];
  }
}
async function handlePythonRepl(code, sessionId, deps) {
  try {
    return await deps.runtime.runRepl({ code, sessionId });
  } catch (err) {
    if (err instanceof PyodideUnavailableError) {
      return {
        stdout: new Uint8Array(),
        stderr: new TextEncoder().encode(`python3: ${err.message}
`),
        exitCode: 127,
        status: "complete"
      };
    }
    const msg = err instanceof Error ? err.message : String(err);
    return {
      stdout: new Uint8Array(),
      stderr: new TextEncoder().encode(`python3: ${msg}
`),
      exitCode: 1,
      status: "complete"
    };
  }
}

export {
  MountMode,
  ConsistencyPolicy,
  ResourceName,
  DEFAULT_SESSION_ID,
  DEFAULT_AGENT_ID,
  FileType,
  FileStat,
  PathSpec,
  CachableAsyncIterator,
  materialize,
  IOResult,
  ExecutionNode,
  ExecutionRecord,
  PyodideUnavailableError,
  handlePython,
  handlePythonRepl
};
